client/asset/btc/spv_scan_filters.go

```diff
@@ -0,0 +1,448 @@
+package btc
+
+import (
+	"bytes"
+	"fmt"
+	"sort"
+	"sync"
+	"time"
+
+	"decred.org/dcrdex/client/asset"
+	"decred.org/dcrdex/dex"
+	"github.com/btcsuite/btcd/chaincfg/chainhash"
+	"github.com/btcsuite/btcd/wire"
+)
+
+// spendingInput is added to a filterScanResult if a spending input is found.
+type spendingInput struct {
+	txHash      chainhash.Hash
+	vin         uint32
+	blockHash   chainhash.Hash
+	blockHeight uint32
+}
+
+// filterScanResult is the result from a filter scan.
+type filterScanResult struct {
+	// blockHash is the block that the output was found in.
+	blockHash *chainhash.Hash
+	// blockHeight is the height of the block that the output was found in.
+	blockHeight uint32
+	// txOut is the output itself.
+	txOut *wire.TxOut
+	// spend will be set if a spending input is found.
+	spend *spendingInput
+	// checkpoint is used to track the last block scanned so that future scans
+	// can skip scanned blocks.
+	checkpoint chainhash.Hash
```

Comment:
Is the checkpoint working correctly? I restarted dexc in the middle of a trade, and it looks like it got in a weird state, but it is checking the same block 717 over and over:

<details closed>
<summary>client log</summary>

```
2024-01-23 10:02:17.879 [DBG] CORE[btc][SPV]: Performing cfilters scan for ee09ed47d28c70eac949a9802b3be83e975758d5883a33ef81aa6e014b5226ab:0 from height 717
2024-01-23 10:02:17.879 [TRC] CORE[btc][SPV]: swapConfirmations - scanFilters did not find ee09ed47d28c70eac949a9802b3be83e975758d5883a33ef81aa6e014b5226ab:0, assuming in mempool.
2024-01-23 10:02:17.884 [ERR] CORE: notify: |ERROR| (order) Swap reporting error - Error notifying DEX of swap for match 5e8c3690cab3ced5787873be196207779d6ca35c2546a2a110cc002b5c811a68: error sending 'init' message: rpc error: error code 64: already received a swap contract, search in progress - Order: 0a3f7c15fcf54c60c51c00dcafcc37e110d0b32f68f60ccb99af9332bba3f13d
2024-01-23 10:02:17.952 [ERR] CORE[wss://127.0.0.1:17273/ws]: No handler found for response: {"type":2,"id":19,"payload":{"result":{"matchid":"5e8c3690cab3ced5787873be196207779d6ca35c2546a2a110cc002b5c811a68","sig":"30440220772a13ba3cfacc400410f7a08cef2ca16fc77a92bd41cd6fe94de93ec9c7e53302207dc64dfa5b0e24deafbee082cb476b59fef7160e2ce1a0a7560931cc28306b34"}},"sig":""}
2024-01-23 10:02:22.221 [INF] CORE: Notifying DEX 127.0.0.1:17273 of our btc swap contract ee09ed47d28c70eac949a9802b3be83e975758d5883a33ef81aa6e014b5226ab:0 for match 5e8c3690cab3ced5787873be196207779d6ca35c2546a2a110cc002b5c811a68
2024-01-23 10:02:22.221 [TRC] CORE[btc][SPV]: swapConfirmations - SpentStatusUnknown: ee09ed47d28c70eac949a9802b3be83e975758d5883a33ef81aa6e014b5226ab:0 (block <nil>, confs 0)
2024-01-23 10:02:22.221 [TRC] CORE[btc][SPV]: swapConfirmations - scanFilters: ee09ed47d28c70eac949a9802b3be83e975758d5883a33ef81aa6e014b5226ab:0 (block <nil>, start time 2024-01-23 19:01:45 +0900 JST)
2024-01-23 10:02:22.221 [DBG] CORE[btc][SPV]: Performing cfilters scan for ee09ed47d28c70eac949a9802b3be83e975758d5883a33ef81aa6e014b5226ab:0 from height 717
2024-01-23 10:02:22.221 [TRC] CORE[btc][SPV]: swapConfirmations - scanFilters did not find ee09ed47d28c70eac949a9802b3be83e975758d5883a33ef81aa6e014b5226ab:0, assuming in mempool.
2024-01-23 10:02:22.221 [TRC] CORE[btc][SPV]: swapConfirmations - SpentStatusUnknown: ee09ed47d28c70eac949a9802b3be83e975758d5883a33ef81aa6e014b5226ab:0 (block <nil>, confs 0)
2024-01-23 10:02:22.221 [TRC] CORE[btc][SPV]: swapConfirmations - scanFilters: ee09ed47d28c70eac949a9802b3be83e975758d5883a33ef81aa6e014b5226ab:0 (block <nil>, start time 2024-01-23 19:01:45 +0900 JST)
2024-01-23 10:02:22.222 [DBG] CORE[btc][SPV]: Performing cfilters scan for ee09ed47d28c70eac949a9802b3be83e975758d5883a33ef81aa6e014b5226ab:0 from height 717
2024-01-23 10:02:22.222 [TRC] CORE[btc][SPV]: swapConfirmations - scanFilters did not find ee09ed47d28c70eac949a9802b3be83e975758d5883a33ef81aa6e014b5226ab:0, assuming in mempool.
2024-01-23 10:02:22.222 [ERR] CORE: Starting match status resolution for 'init' request SettlementSequenceError
2024-01-23 10:02:22.225 [ERR] CORE: notify: |ERROR| (order) Swap reporting error - Error notifying DEX of swap for match 5e8c3690cab3ced5787873be196207779d6ca35c2546a2a110cc002b5c811a68: error sending 'init' message: rpc error: error code 26: expected other party to act - Order: 0a3f7c15fcf54c60c51c00dcafcc37e110d0b32f68f60ccb99af9332bba3f13d
2024-01-23 10:02:24.449 [DBG] CORE[btc]: tip change: 716 (73a57a800a25d59cccba030798e717c060ec927b8dce7a91ac56e38fa8cb9baf) => 717 (78b77f4a180fe5a380a7f46503cd0f58c5c7f319bc8d341659211f0f28ab5ebf)
2024-01-23 10:02:24.449 [TRC] CORE: Processing tip change for btc
2024-01-23 10:02:24.449 [INF] CORE: Notifying DEX 127.0.0.1:17273 of our btc swap contract ee09ed47d28c70eac949a9802b3be83e975758d5883a33ef81aa6e014b5226ab:0 for match 5e8c3690cab3ced5787873be196207779d6ca35c2546a2a110cc002b5c811a68
2024-01-23 10:02:24.449 [TRC] CORE[btc][SPV]: swapConfirmations - SpentStatusUnknown: ee09ed47d28c70eac949a9802b3be83e975758d5883a33ef81aa6e014b5226ab:0 (block 78b77f4a180fe5a380a7f46503cd0f58c5c7f319bc8d341659211f0f28ab5ebf, confs 1)
2024-01-23 10:02:24.449 [TRC] CORE[btc][SPV]: swapConfirmations - scanFilters: ee09ed47d28c70eac949a9802b3be83e975758d5883a33ef81aa6e014b5226ab:0 (block 78b77f4a180fe5a380a7f46503cd0f58c5c7f319bc8d341659211f0f28ab5ebf, start time 2024-01-23 19:01:45 +0900 JST)
2024-01-23 10:02:24.449 [DBG] CORE[btc][SPV]: Performing cfilters scan for ee09ed47d28c70eac949a9802b3be83e975758d5883a33ef81aa6e014b5226ab:0 from height 717
2024-01-23 10:02:24.449 [TRC] CORE[btc][SPV]: Block 78b77f4a180fe5a380a7f46503cd0f58c5c7f319bc8d341659211f0f28ab5ebf matched pkScript for output ee09ed47d28c70eac949a9802b3be83e975758d5883a33ef81aa6e014b5226ab:0. Pulling the block...
2024-01-23 10:02:24.449 [TRC] CORE[btc][SPV]: Found txn ee09ed47d28c70eac949a9802b3be83e975758d5883a33ef81aa6e014b5226ab in block 78b77f4a180fe5a380a7f46503cd0f58c5c7f319bc8d341659211f0f28ab5ebf (717)
2024-01-23 10:02:24.449 [DBG] CORE[btc][SPV]: cfilters scan SUCCEEDED for ee09ed47d28c70eac949a9802b3be83e975758d5883a33ef81aa6e014b5226ab:0. block hash: 78b77f4a180fe5a380a7f46503cd0f58c5c7f319bc8d341659211f0f28ab5ebf, spent: false
2024-01-23 10:02:24.449 [TRC] CORE[btc][SPV]: swapConfirmations - SpentStatusUnknown: ee09ed47d28c70eac949a9802b3be83e975758d5883a33ef81aa6e014b5226ab:0 (block 78b77f4a180fe5a380a7f46503cd0f58c5c7f319bc8d341659211f0f28ab5ebf, confs 1)
2024-01-23 10:02:24.449 [TRC] CORE[btc][SPV]: swapConfirmations - scanFilters: ee09ed47d28c70eac949a9802b3be83e975758d5883a33ef81aa6e014b5226ab:0 (block 78b77f4a180fe5a380a7f46503cd0f58c5c7f319bc8d341659211f0f28ab5ebf, start time 2024-01-23 19:01:45 +0900 JST)
2024-01-23 10:02:24.449 [DBG] CORE[btc][SPV]: Performing cfilters scan for ee09ed47d28c70eac949a9802b3be83e975758d5883a33ef81aa6e014b5226ab:0 from height 718
```
</details>

Reply:
I think it is working as expected. The initial repeated block 717 checks was because the tip was 716. So it had probably checked up till 716 and didn't find it. It was not able to check 717 because there's no 717 yet, so each attempt kept trying to start the search from 717. Until the tip change to 717 happened and then it started the search from 717 again and found the tx this time.

The next scan started from 718 (not 717 again); and this is likely the search for the spender.